terraform/lib/dependabot/terraform/registry_client.rb

```diff
@@ -98,6 +81,47 @@ def headers_for(hostname)
         token = tokens[hostname]
         token ? { "Authorization" => "Bearer #{token}" } : {}
       end
+
+      def services
+        @services ||=
+          begin
+            response = http_get(url_for("/.well-known/terraform.json"))
+            response.status == 200 ? JSON.parse(response.body) : {}
+          end
+      end
+
+      def service_key_for(name)
+        case name
+        when "module", "modules", "registry"
+          "modules.v1"
+        when "provider", "providers"
+          "providers.v1"
+        else
+          raise "Invalid source type"
+        end
+      end
+
+      def service_url_for(service_key)
+        url_for(services.fetch(service_key))
+      rescue KeyError
+        raise Dependabot::DependabotError, "Host does not support required Terraform-native service"
+      end
+
+      def http_get(url)
+        Excon.get(url.to_s, idempotent: true, **SharedHelpers.excon_defaults(headers: headers_for(hostname)))
+      end
+
+      def http_get!(url)
+        response = http_get(url)
+
+        raise "Response from registry was #{response.status}" unless response.status == 200
+
+        response
+      end
+
+      def url_for(path)
+        "https://#{hostname}#{path}"
```

Comment:
Question: should this `URI.join().to_s`? I'm paranoid about trailing slashes.

Reply:
Concern: the spec says `The value of each object element is the base URL for the service in question. This URL may be either absolute or relative, and if relative it is resolved against the final discovery URL (after following redirects).`

I'm concerned about this implementation of `service_url_for()` when given an absolute URL response:

```json
{
  "modules.v1": "https://modules.intranet.example.com/v1/"
}
```

I think it would return something like `"https://modules.example.comhttps://modules.intranet.example.com/v1/"`

Reply:
Thanks Pete. I added some tests and made an adjustment with how we compose the service url so that is handles both relative and absolute urls.